### DISCUSSION TITLE
LAZY_HOME to home Z in the middle of the bed?

### DISCUSSION THREAD

**Discussion by academo** [jschuh/klipper-macros#139] [Q&A] [answered]
I like the concept of LAZY_HOME but I can't figure how to tell it to home z in the middle of the bed since I have a probe. When I first tried to use it, it tried to home z in a spot where the probe won't hit the bed.

I fixed this by writing my own home_overwrite with `G28.6245197` as suggested in the readme.

Is there a way to instruct LAZY_HOME to move to the center of the bed before trying to home?

Thanks for the work in these scripts.

**Comment by jschuh** [comment, marked as answer]
You're doing it exactly the right way. Changes to core homing functionality are really best handled with either `homing_override` or `safe_z_home` (depending on your needs). So, it doesn't make sense to duplicate that functionality in a macro.

**Reply by academo** [reply]
Ah! makes sense. This is what I have in my home_override which is already lazy-homing. Should I remove all this lazy-home logic and simply home when requested? Is lazy_home called when G28 is normally called?

```
[homing_override]
axes: xyz
gcode:
    M118 Homing Script starteds
    {% set is_homed_x = False %}
    {% set is_homed_y = False %}
    {% set is_homed_z = False %}

    {% set should_home_X = False %}
    {% set should_home_Y = False %}
    {% set should_home_Z = False %}

    {% if "x" in printer.toolhead.homed_axes %} {% set is_homed_x = True %} {% endif %}
    {% if "y" in printer.toolhead.homed_axes %} {% set is_homed_y = True %} {% endif %}
    {% if "z" in printer.toolhead.homed_axes %} {% set is_homed_z = True %} {% endif %}

    {% if "x" in params|map('lower') %} {% set should_home_X = True %} {% endif %}
    {% if "y" in params|map('lower') %} {% set should_home_Y = True %} {% endif %}
    {% if "z" in params|map('lower') %} {% set should_home_Z = True %} {% endif %}

    ; G28 command alone
    {% if not should_home_X and not should_home_Y and not should_home_Z %}
        {% set should_home_X = True %}
        {% set should_home_Y = True %}
        {% set should_home_Z = True %}
    {% endif %}

    # M118 params {params}
    # M118 axes status {printer.toolhead.homed_axes}
    # M118 is_homed_x {is_homed_x}
    # M118 is_homed_y {is_homed_y}
    # M118 is_homed_z {is_homed_z}
    # M118 should_home_X {should_home_X}
    # M118 should_home_Y {should_home_Y}
    # M118 should_home_Z {should_home_Z}

    {% if should_home_X and not is_homed_x %}.
        M118 Homing X
        G28.6245197 X
        {% set is_homed_x = True %}
    {% else %}
        M118 X already homed. Skipping
    {% endif %}

    {% if should_home_Y and not is_homed_y %}
        M118 Homing Y
        G28.6245197 Y
        {% set is_homed_y = True %}
    {% else %}
        M118 Y already homed. Skipping
    {% endif %}

    {% if should_home_Z and not is_homed_z %}
        M118 Homing Z (checks)
        {% if not is_homed_x %}
            M118 Homing X
            G28.6245197 X
        {% endif %}
        {% if not is_homed_y %}
            M118 Homing Y
            G28.6245197 Y
        {% endif %}
        M118 Homing Z
        SET_KINEMATIC_POSITION z=0
        G1 Z10 F600
        G1 X170 Y120 F6000
        G28.6245197 Z
        {% set is_homed_z = True %}
    {% else %}
        M118 Z already homed. Skipping
    {% endif %}
    M118 Homing Script Finished```

**Reply by jschuh** [reply]
Yes, `homing_override` replaces the built-in G28 command, so LAZY_HOME (or anything else that calls the built-in G28) will end up calling the `homing_override` gcode. That's why it's illegal to call a wrapped G28 inside of `homing_override`, because that would cause a recursive macro call.

You're correct about duplicating the LAZY_HOME logic. I do lazily home the X and Y in my `homing_override`, but only when the homing request is solely for the Z axis. My rationale is that I prefer lazy homing to be explicitly requested, rather than an implicit default (as it is in the gcode you provided).

Here's a heavily commented version of the `homing_override` that I use for probes (with your offsets plugged in):

```
[homing_override]
set_position_z: 0
axes: z
gcode:
  G90 # Absolute position mode
  G1 Z10 F600 # Move up a safe distance
  # Ensure X and Y are homed.
  {% if 'x' not in printer.toolhead.homed_axes %}G28.6245197 X{% endif %}
  {% if 'y' not in printer.toolhead.homed_axes %}G28.6245197 Y{% endif %}
  G1 X170 Y120 F6000 # Move probe to center
  G28.6245197 Z
```

**Reply by academo** [reply]
Thanks for the explanation!
